### PR TITLE
Using ``serialize_method`` instead of astropy_native for writing

### PR DESCRIPTION
**Pull request by AustereCuriosity**
``time_col.info.serialize_method['fits'] == 'jd1_jd2'`` is the default, due to which, for writing of Time columns, the default behaviour is full precision (jd1, jd2) format. For writing the actual representation values of the time columns, we explicitly set ``time_col.info.serialize_method['fits'] = 'formatted_value'`` to get the formatted value.

@taldcroft @mhvk @hamogu

### CODE REVIEW

File: astropy/io/fits/tests/test_fitstime.py [hunk 1]
```diff
@@ -129,6 +129,7 @@ def test_fits_to_time_meta(self, table_types):
         t.meta['MJD-OBS'] = 56670
 
         # Test for default read/write behaviour (raw data)
```
Comment (mhvk) [inline astropy/io/fits/tests/test_fitstime.py]: This comment should be adjusted. Maybe just "Test for conversion of time data to its `value`, as defined by its format" or something like that.

As it is *not* the default, I would also suggest to swap the test, so that the default is tested first (but this is really a very minor detail).

File: astropy/io/fits/tests/test_fitstime.py [hunk 2]
```diff
@@ -141,8 +142,8 @@ def test_fits_to_time_meta(self, table_types):
         assert tm.meta['MJD-OBS'] == t.meta['MJD-OBS']
 
         # Test for native astropy objects read/write behaviour
```
Comment (mhvk) [inline astropy/io/fits/tests/test_fitstime.py]: Similarly, this *can* now be called the default.

File: astropy/table/tests/test_mixin.py [hunk 3]
```diff
@@ -185,6 +185,11 @@ def test_io_time_write_fits(tmpdir, table_types):
         assert (tm[name] == t[name]).all()
 
     # Test for default read/write behaviour (raw data)
```
Comment (mhvk) [inline astropy/table/tests/test_mixin.py]: Same - this is not the default.

File: astropy/io/fits/fitstime.py [hunk 4]
```diff
@@ -355,6 +355,10 @@ def time_to_fits(table):
     location = None
 
     for col in time_cols:
+        if col.info.serialize_method['fits'] == 'formatted_value':
```
Comment (mhvk) [inline astropy/io/fits/fitstime.py]: Could you add a code comment here, along the lines of: "by default, we write time objects in full precision, i.e., we store both jd1 and jd2 (serialize_method['fits'] = 'jd1_jd2'), unless the user explicitly chose to store the formatted value." (not great text, do adjust!)

**Comment by AustereCuriosity** [comment]
@mhvk Yes, I have mostly changed the documentation to indicate these changes. I'll push it here asap.

**Comment by AustereCuriosity** [comment]
This is ready for a final review. @mhvk @taldcroft @hamogu

File: astropy/io/fits/tests/test_fitstime.py [hunk 5]
```diff
@@ -177,6 +164,19 @@ def test_fits_to_time_meta(self, table_types):
...
+        tm = table_types.read(self.temp('time.fits'), format='fits')
+
+        # Test DATE
+        assert not isinstance(tm.meta['DATE'], Time)
+        assert tm.meta['DATE'] == t.meta['DATE']
+
+        # Test MJD-xxx
+        assert not isinstance(tm.meta['MJD-OBS'], Time)
+        assert tm.meta['MJD-OBS'] == t.meta['MJD-OBS']
+
```
Comment (mhvk) [inline astropy/io/fits/tests/test_fitstime.py]: I think there should be an additional test that `np.all(tn['a'] == t['a'].value)` (since the time has format `isot`, `.value` is a string, and so in comparing there are no worries about rounding errors etc.)

File: docs/io/unified.rst [hunk 6]
```diff
@@ -279,28 +279,26 @@ Time
...
-console).  For instance, if the ``format`` is ``iso`` then the FITS column will
-be a character string column.  None of the special header keywords associated
-with the Time FITS standard will be set.  When reading this back into Astropy,
-the column will be an ordinary Column instead of a `~astropy.time.Time` object.
-See the `Details`_ section below for an example.
+or `~astropy.table.QTable` will be written to FITS in full precision. This will be
+done using the FITS Time standard by setting the necessary FITS header keywords.
+
+The default read behaviour of a FITS table into an Astropy Table is to read the FITS
+table columns as standard Table Columns. The ``astropy_native``keyword has thus been
```
Comment (mhvk) [inline docs/io/unified.rst]: missing space after the closing double back-tick.

File: docs/io/unified.rst [hunk 7]
```diff
@@ -279,28 +279,26 @@ Time
...
-the ``format`` of the object (similar to the output when printing to the
-console).  For instance, if the ``format`` is ``iso`` then the FITS column will
-be a character string column.  None of the special header keywords associated
-with the Time FITS standard will be set.  When reading this back into Astropy,
-the column will be an ordinary Column instead of a `~astropy.time.Time` object.
-See the `Details`_ section below for an example.
+or `~astropy.table.QTable` will be written to FITS in full precision. This will be
+done using the FITS Time standard by setting the necessary FITS header keywords.
+
+The default read behaviour of a FITS table into an Astropy Table is to read the FITS
```
Comment (mhvk) [inline docs/io/unified.rst]: Maybe rephrase a bit:
```
The default behaviour for reading a FITS table into an `~astropy.table.Table`
has historically been to convert all FITS columns to `~astropy.table.Column`
objects, which have closely matching properties. For some columns, however,
closer native astropy representations are possible, and one can indicate these
should be used by passing ``astropy_native=True`` (for backwards compatibility,
this is not done by default). This will convert columns conforming to the
FITS time standard to `~astropy.time.Time` instances, avoiding any loss of
precision. For example::
```

File: docs/io/unified.rst [hunk 8]
```diff
@@ -318,6 +316,15 @@ the keywords ``DATE``, ``DATE-*`` (ISO-8601 datetime strings) and the ``MJD-*``
 For more details regarding the FITS time paper and the implementation,
```
Comment (mhvk) [inline docs/io/unified.rst]: This is really a comment for line 310 (for `QTable`). I think it should have a bit of extra text, as follows. Given that, I also think it makes sense to have ``astropy_native=True`` as the default for `QTable`...
```
Note that even with ``astropy_native=True``, numerical columns with units will
be stored as `~astropy.table.Column` instances, i.e., they will not be
converted to `~astropy.units.Quantity` objects. For that, one should use
`~astropy.table.QTable`; for that class, ``astropy_native`` is set to `True`
by default.
```

File: docs/io/unified.rst [hunk 9]
```diff
@@ -318,6 +316,15 @@ the keywords ``DATE``, ``DATE-*`` (ISO-8601 datetime strings) and the ``MJD-*``
...
 refer to :ref:`fits_time_column`.
 
+As opposed to the default write behaviour of a `~astropy.time.Time` mixin column
```
Comment (mhvk) [inline docs/io/unified.rst]: I'd again rephrase the rationale (linking directly to the Time documentation):
```
Since not all FITS readers are able to use the FITS time standard, it is also
possible to store `~astropy.time.Time` instances using the `_time_format`.
For this case, none of the ...
```

File: docs/io/unified.rst [hunk 10]
```diff
@@ -352,23 +359,24 @@ the final conversion from ``(jd1, jd2)`` requires a software implementation whic
 fully compliant with the FITS Time standard.
 
 Taking this into consideration, the functionality to read/write Time
-from/to FITS is off by default. Users can store the time
+from/to FITS can be explicitly chosen to be off. Users can store the time
```
Comment (mhvk) [inline docs/io/unified.rst]: ``can be explicitly turned off, by opting to store the time``

File: docs/io/unified.rst [hunk 11]
```diff
@@ -352,23 +359,24 @@ the final conversion from ``(jd1, jd2)`` requires a software implementation whic
...
 representation values in the format specified by the ``format`` attribute
 of the `~astropy.time.Time` column, instead of the ``(jd1, jd2)`` format, with
-no extra metadata in the header. This is the "lossy" version,
-but meets common-use needs. For the above example, the FITS column corresponding
+no extra metadata in the header. This is the "lossy" version, but meets
```
Comment (mhvk) [inline docs/io/unified.rst]: Instead of ``but meets common-use needs``, maybe ``but can help portability``

File: docs/io/unified.rst [hunk 12]
```diff
@@ -352,23 +359,24 @@ the final conversion from ``(jd1, jd2)`` requires a software implementation whic
...
+common-use needs. For the above example, the FITS column corresponding
 to ``t['a']`` will then store ``[100.0 200.0]`` instead of
-``[[ 2400100.5, 0. ], [ 2400200.5, 0. ]]``.
-This will be useful when dealing with other software packages which do not implement
-the full FITS time standard or when users want to serialize their Time object
-in the format of their choosing.
+``[[ 2400100.5, 0. ], [ 2400200.5, 0. ]]``, and this is done by setting
```
Comment (mhvk) [inline docs/io/unified.rst]: I think I would just write ``. This is done by using a special ``info.serialze_method`` attribute, as in the following example::`` (@taldcroft - I was going to suggest to put a link to the `ecsv` documentation about the similar use there, but couldn't find it - did we actually document this?)

File: docs/io/unified.rst [hunk 13]
```diff
@@ -378,6 +386,13 @@ in the format of their choosing.
     >>> tm['a'] == t['a'].value
     array([ True,  True], dtype=bool)
 
+This write behaviour of Time will be useful when dealing with other software packages
```
Comment (mhvk) [inline docs/io/unified.rst]: I think you just explained that above, so would suggest deleting this paragraph.

**Comment by hamogu** [comment]
We discussed defaults in one of our telecons. I agree with @mhvk that it makes sense to have `QTable` do everything as astropy-like as possible, but I think @taldcroft was opposed because he fears that users get confused. Currently, there is one and only one difference between `Table` and `QTable` - columns are quantities. With this, we start creeping in more differences and thus in the written docs we'll always have to say "Here is an example for Table, but note that QTable has the opposite default", "This is how you change the the parameter to get the old behavior, but notice the this is only necessary for QTable, while in Table it is exactly the other way around", etc.

**Comment by mhvk** [comment]
OK, I see that point, although I think there is also a logic to `QTable` having "native whenever possible" as its defining feature (I'd be happy for it not even allowing ``astropy_native=False``, i.e., not having the keyword at all).

But I definitely don't want to distract from this PR - this should get in regardless and I think in the end we can handle these matters of taste also via a config item.

File: docs/io/unified.rst (hunk 8 above)
Comment (mhvk) [inline docs/io/unified.rst]: @AustereCuriosity - given the comments in the main thread, maybe the last part should be
```
`~astropy.table.QTable` (but otherwise, the classes behave identically).
```

**Comment by AustereCuriosity** [comment]
@mhvk @taldcroft @hamogu This is ready for merging.

File: docs/io/unified.rst (hunk 8 above)
Comment (taldcroft) [inline docs/io/unified.rst]: I disagree about changing the `astropy_native` default for `QTable`:
- It is an API change. If someone is currently reading FITS tables with time columns or date keywords via `QTable`, those will suddenly turn into Time objects.  One should not assume that the only people using QTable are bleeding edge devs that are fine with code breaking.
- The distinction between Table and QTable is already confusing enough.  I have worked *really* hard to emphasize that "the ONLY, ONLY difference between Table and QTable is in the handling of columns with a unit".  Then I repeat that 5 times.  So now we want to add a footnote that, "also, it reads FITS files differently by default".

File: docs/io/unified.rst [hunk 14]
```diff
@@ -349,26 +356,25 @@ the desired format. In the example stated above, the Time column ``t['a']``
...
-in the format of their choosing.
+``[[ 2400100.5, 0. ], [ 2400200.5, 0. ]]``. This is done by using a special
+``info.serialize_method`` attribute, as in the following example::
 
     >>> from astropy.time import Time
     >>> from astropy.table import Table
     >>> from astropy.coordinates import EarthLocation
     >>> t = Table()
     >>> t['a'] = Time([100.0, 200.0], scale='tt', format='mjd',
     ...               location=EarthLocation(-2446354, 4237210, 4077985, unit='m'))
```
Comment (taldcroft) [inline docs/io/unified.rst]: This `location` is irrelevant, so you can cut it out.

File: docs/io/unified.rst [hunk 15]
```diff
@@ -309,7 +306,10 @@ round-tripping the object with no loss of precision.  For example::
     >>> tm['a'] == t['a']
     array([ True,  True], dtype=bool)
 
-The same will work with ``QTable``.
+Note that even with ``astropy_native=True``, numerical columns with units will
```
Comment (taldcroft) [inline docs/io/unified.rst]: Somehow this new text implies that `Column` is not an astropy native object.  I think that this whole paragraph is confusing in this spot because it is a hard segway from `Time` to a somewhat unrelated topic of `QTable` vs. `Table` handling of `Quantity`.  I liked the original "The same will work with QTable".

But if we really want this paragraph then change `stored` to `read`, because `stored` implies that the sentence is related to writing, not reading.  (I don't think of "storing" columns into a table in memory).

File: docs/io/unified.rst (hunk 14 above)
Comment (AustereCuriosity) [inline docs/io/unified.rst]: I've changed it. I have no clue why isn't showing this as changed.

Comment (taldcroft) [inline docs/io/unified.rst]: Well don't worry about it.

**Comment by AustereCuriosity** [comment]
@taldcroft @mhvk @hamogu This is ready to be merged in. The two failing jobs are unrelated (urlopen errors).

**Comment by AustereCuriosity** [comment]
@taldcroft @mhvk @hamogu Just a reminder that this is still open.
